task (#516) : update derivation and transaction internalization methods

cain4chain opened the pull request:
## Description of Changes

- Derivation method : replace self counterparty method with anyoneKey value for key derivations in wallet examples.
- Transaction Internalization : update transaction internalization method to allow output matching validation.
- Documentation : update documentation based on the changes.

## Linked Issues / Tickets

Closes #516 

## Testing Procedure

Describe the tests you've added or any testing steps you've taken.

- [ ] I have added new unit tests
- [ ] All tests pass locally
- [ ] I have tested manually in my local environment

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have run the linter

cain4chain committed:
task : update derivation and transaction internalization methods

mgosek-4chain commented on examples/internal/example_setup/faucet_address.go:
```diff
@@ -18,8 +19,9 @@ func FaucetAddress(wallet *Setup) {
 		DerivationSuffix: base64.StdEncoding.EncodeToString(parts.DerivationSuffix),
 	}
 
+	anyonePriv, _ := sdk.AnyoneKey()
```
Please add basic error handling :)

chris-4chain replied:
This is not an error 😅 The function returns priv and pub keys.

dorzepowski commented on examples/wallet_examples/internalize_wallet_payment/internalize_wallet_payment.go:
```diff
@@ -62,31 +70,77 @@ func main() {
+	if err != nil {
+		panic(fmt.Errorf("failed to create locking script for address %q: %w", derivedAddress, err))
+	}
+
```
hmmm... It looks very cool, BUT
not for this example:
The wallet examples shouldn't be very smart, but show how to use wallet methods in a simple way.
Therefore I would rather suggest to 
1. use this smart searching in internalize from faucet example
2. revert the changes in this file
3. add const with vout in this file and use it in args (like it is done for other parameters)

dorzepowski commented on examples/wallet_examples/internalize_wallet_payment/internalize_wallet_payment.go:
```diff
@@ -62,31 +70,77 @@ func main() {
+	addressObj, err := script.NewAddressFromString(derivedAddress.AddressString)
+	if err != nil {
+		panic(fmt.Errorf("failed to parse address %q: %w", derivedAddress, err))
+	}
```
why you're not using brc29.Address function to get the address?
Is this code doing deriving some other address then the brc29 address?

cain4chain replied:
Yes in this case we are only providing the senderIdentityKey as the counterparty in the derivation process. Since we assume the transaction that is coming from any specified party we cannot reuse the private key that may have derived the original address.

cain4chain committed:
refactor : Update internalization examples

- Revert changes to internalize wallet payment example, add outputIndex parameter for specified output to be internalized
- Add outputIndex validation to internalize faucet tx example
- Update documentation

cain4chain committed:
Merge branch 'main' into task-wallet-examples---update-derivation-and-transaction-internalization-methods

chris-4chain reviewed (commented):
This PR is potentially conflicting with this one: https://github.com/bsv-blockchain/go-wallet-toolbox/pull/520 (by @dorzepowski)

Please make sure you are one the same page
